numpyro/util.py

```diff
@@ -163,6 +165,120 @@ def _body_fn(state):
+                   diag_mass=True, target_accept_prob=0.8):
+    ss_init, ss_update = dual_averaging()
+    mm_init, mm_update, mm_final = welford_covariance(diagonal=diag_mass)
+    adaptation_schedule = np.array(build_adaptation_schedule(num_steps))
+    num_windows = len(adaptation_schedule)
+
+    def init_fn(step_size=1.0, inverse_mass_matrix=None, mass_matrix_size=None):
+        if find_reasonable_step_size is not None:
+            step_size = find_reasonable_step_size(step_size)
+        ss_state = ss_init(np.log(10 * step_size))
+
+        if inverse_mass_matrix is None:
+            assert mass_matrix_size is not None
+            if diag_mass:
+                inverse_mass_matrix = np.ones(mass_matrix_size)
+            else:
+                inverse_mass_matrix = np.identity(mass_matrix_size)
+        mm_state = mm_init(inverse_mass_matrix.shape[-1])
+
+        window_idx = 0
+        return step_size, inverse_mass_matrix, ss_state, mm_state, window_idx
+
+    def _update_at_window_end(state):
+        step_size, inverse_mass_matrix, ss_state, mm_state, window_idx = state
+
+        if adapt_step_size:
+            if find_reasonable_step_size is not None:
+                step_size = find_reasonable_step_size(step_size)
+            ss_state = ss_init(np.log(10 * step_size))
+
+        if adapt_mass_matrix:
+            inverse_mass_matrix = mm_final(mm_state, regularize=True)
+            mm_state = mm_init(inverse_mass_matrix.shape[-1])
+
+        return step_size, inverse_mass_matrix, ss_state, mm_state, window_idx
+
+    def update_fn(t, accept_prob, z_flat, state):
+        step_size, inverse_mass_matrix, ss_state, mm_state, window_idx = state
+
+        # update step size state
+        if adapt_step_size:
+            ss_state = ss_update(target_accept_prob - accept_prob, ss_state)
+            # note: at the end of warmup phase, use average of log step_size
+            # TODO: should we make sure that we won't update step_size if t >= num_steps?
+            log_step_size, log_step_size_avg, *_ = ss_state
+            step_size = np.where(t == (num_steps - 1),
+                                 np.exp(log_step_size_avg),
+                                 np.exp(log_step_size))
+
+        # update mass matrix state
+        is_middle_window = (0 < window_idx) & (window_idx < (num_windows - 1))
+        if adapt_mass_matrix:
+            mm_state = lax.cond(is_middle_window,
+                                (z_flat, mm_state), lambda args: mm_update(*args),
+                                mm_state, lambda x: x)
+
+        t_at_window_end = t == adaptation_schedule[window_idx, 1]
+        window_idx = np.where(t_at_window_end, window_idx + 1, window_idx)
+        state = step_size, inverse_mass_matrix, ss_state, mm_state, window_idx
+        # TODO: enable lax.cond when https://github.com/google/jax/issues/514 is resolved
+        # state = lax.cond(t_at_window_end & is_middle_window,
```

Comment:
If you didn't pass in `inverse_mass_matrix` and `step_size` i.e. the `_update_at_window_end` only took 2 arguments (I know that wouldn't work when mass matrix or step size adaptation is disabled), would the test pass with jit?

Reply:
@neerajprad I just test it. JIT works well given that prox_center of `ss_init` is a constant (not depends on current `ss_state`).

Reply:
@neerajprad FYI, the fix is available in JAX master branch and the test with all states passes.